File: OrchidCore/src/main/java/com/eden/orchid/api/theme/assets/DefaultAssetManager.java
package com.eden.orchid.api.theme.assets;

import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.resources.resource.OrchidResource;
import com.eden.orchid.impl.themes.functions.ThumbnailResource;
import javax.inject.Provider;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.HashMap;
import java.util.Map;

@Singleton
public final class DefaultAssetManager implements AssetManager {

    private final Provider<OrchidContext> context;

    private final Map<String, AssetPage> assets;

    @Inject
    public DefaultAssetManager(Provider<OrchidContext> context) {
        this.context = context;
        this.assets = new HashMap<>();
    }

    @Override
    public AssetPage createAsset(String asset, Object source, String sourceKey) {
        OrchidResource originalResource = context.get().getResourceEntry(asset);
        if (originalResource != null) {
            // don't render the asset immediately. Allow the template to apply transformations to the asset, and it will be
            // rendered lazily when the link for the asset is requested (or not at all if it is never used)
            return new AssetPage(source, sourceKey, new ThumbnailResource(originalResource), "thumbnail", originalResource.getTitle());
        }

        return null;
    }

    @Override
    public AssetPage getActualAsset(AssetPage asset) {
        final String assetKey = asset.getReference().toString();
        if(assets.containsKey(assetKey)) {
            return assets.get(assetKey);
        }
        else {
            return null;
        }
    }

    // Assets should only make it here if it passes the check in a local AssetHolderDelegate, so we don't need to check
    // again. Go ahead and render it now so we can free its resources, and eventually implement an asset pipeline from
    // this point. Inline resources are rendered directly into the page, and should not be rendered as a proper resource
    @Override
    public AssetPage addAsset(AssetPage asset, boolean renderImmediately) {
        final String assetKey = asset.getReference().toString();
        if(assets.containsKey(assetKey)) {
            return assets.get(assetKey);
        }
        else {
            assets.put(assetKey, asset);
            if(renderImmediately) {
                context.get().renderAsset(asset);
            }

            return asset;
        }
    }

    @Override
    public void clearAssets() {
        assets.clear();
    }



}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/theme/assets/DefaultAssetManager.kt
package com.eden.orchid.api.theme.assets

import com.eden.orchid.api.resources.resource.OrchidResource
import com.eden.orchid.api.resources.resource.StringResource
import com.eden.orchid.api.resources.resource.ThumbnailResource
import com.eden.orchid.api.resources.resourcesource.OrchidResourceSource
import com.eden.orchid.api.theme.pages.OrchidReference
import java.util.stream.Stream
import javax.inject.Singleton

@Singleton
class DefaultAssetManager : AssetManager {

    // TODO: make this a cache, not a map
    // TODO: have the keys be asset hashCodes (not resource paths) for improved performance
    private val assets: MutableMap<String, AssetPage> = mutableMapOf()

    override val allAssetPages: Stream<AssetPage> get() = assets.values.stream()

    override fun clearAssets() {
        assets.clear()
    }

    override fun createAsset(origin: AssetManagerDelegate, resourceSource: OrchidResourceSource, asset: String): AssetPage {
        val requestedResource: OrchidResource? = resourceSource.getResourceEntry(origin.context, asset)

        checkNotNull(requestedResource) { "requested asset '$asset' could not be found" }

        return AssetPage(
            origin,
            ThumbnailResource(requestedResource),
            "thumbnail",
            requestedResource.reference.title
        ).also { it.configureReferences() }
    }

    override fun createCss(origin: AssetManagerDelegate, resourceSource: OrchidResourceSource, asset: String, configure: CssPageAttributes?): CssPage {
        val requestedResource: OrchidResource? = resourceSource.getResourceEntry(origin.context, asset)

        checkNotNull(requestedResource) { "requested asset '$asset' could not be found" }

        return CssPage(
            origin,
            requestedResource,
            "css",
            requestedResource.reference.title
        ).also {
            if(configure != null) {
                it.applyAttributes(configure)
            }
            it.configureReferences()
        }
    }

    override fun createJs(origin: AssetManagerDelegate, resourceSource: OrchidResourceSource, asset: String, configure: JsPageAttributes?): JsPage {
        val requestedResource: OrchidResource? = resourceSource.getResourceEntry(origin.context, asset)

        checkNotNull(requestedResource) { "requested asset '$asset' could not be found" }

        return JsPage(
            origin,
            requestedResource,
            "js",
            requestedResource.reference.title
        ).also {
            if(configure != null) {
                it.applyAttributes(configure)
            }
            it.configureReferences()
        }
    }

    override fun getActualAsset(origin: AssetManagerDelegate, asset: AssetPage, renderImmediately: Boolean): AssetPage {
//        val assetKey = asset.hashCode()
        val assetKey = asset.reference.toString(asset.context)

        if (!assets.containsKey(assetKey)) {
            if (renderImmediately) {
                // internally, this will check if the asset is inlined or not, and set the `isRendered` flag accordingly
                origin.context.renderAsset(asset)
            }
            assets[assetKey] = asset
        }

        return assets[assetKey]!!
    }

    override fun getActualCss(origin: AssetManagerDelegate, asset: CssPage, renderImmediately: Boolean): CssPage {
        val actualAsset = getActualAsset(origin, asset, renderImmediately)

        check(actualAsset is CssPage) { "Cached asset with hashcode ${actualAsset.hashCode()} is not a CssPage" }

        return actualAsset
    }

    override fun getActualJs(origin: AssetManagerDelegate, asset: JsPage, renderImmediately: Boolean): JsPage {
        val actualAsset = getActualAsset(origin, asset, renderImmediately)

        check(actualAsset is JsPage) { "Cached asset with hashcode ${actualAsset.hashCode()} is not a JsPage" }

        return actualAsset
    }
}
